docs/source/main_classes/deepspeed.mdx

```diff
@@ -1751,6 +1751,171 @@ Please note that if you're not using the [`Trainer`] integration, you're complet
 [[autodoc]] deepspeed.HfDeepSpeedConfig
     - all
 
+### DeepSpeed ZeRO Inference
+
+Here is an example of how one could do DeepSpeed ZeRO Inference without using [`Trainer`] when one can't fit a model onto a single GPU. The solution includes using additional GPUs or/and offloading GPU memory to CPU memory.
+
+The important nuance to understand here is that the way ZeRO is designed you can process different inputs on different gpus in parallel.
+
+The example has copious notes and is self-documenting.
+
+Make sure to:
+
+1. disable cpu offload if you have enough GPU memory (since it slows things down)
+2. enable bf16 if you own an Ampere or a newer gpu to make things faster. If you don't have that hardware you may enable fp16 as long as you use a different model than a t5-variant or any other model that was pre-trained in bf16 mixed precision - these usually overflow in fp16 and you will see garbage as output.
+
+```
```

Comment:
Should there be python here?

Reply:
hmm, let me move out the bash call and then yes we can put `python` there.

Reply:
grr, switching to `python` now the autoformatter wants to interfere with the code example :(

is there any way to tell it not to change some parts?

It's trying to change this:

```
@@ -1847,20 +1847,17 @@ ds_config = {
     },
     "zero_optimization": {
         "stage": 3,
-        "offload_param": {
-            "device": "cpu",
-            "pin_memory": True
-        },
+        "offload_param": {"device": "cpu", "pin_memory": True},
         "overlap_comm": True,
         "contiguous_gradients": True,
         "reduce_bucket_size": model_hidden_size * model_hidden_size,
         "stage3_prefetch_bucket_size": 0.9 * model_hidden_size * model_hidden_size,
-        "stage3_param_persistence_threshold": 10 * model_hidden_size
+        "stage3_param_persistence_threshold": 10 * model_hidden_size,
     },
     "steps_per_print": 2000,
     "train_batch_size": train_batch_size,
     "train_micro_batch_size_per_gpu": 1,
-    "wall_clock_breakdown": False
+    "wall_clock_breakdown": False,
```

which makes it much more difficult to copy to/from json file, including collapsing vertical space which makes the config much more difficult to deal with.

Reply:
You can see an example of how json wants it here: https://github.com/huggingface/transformers/blob/master/tests/deepspeed/ds_config_zero3.json

Reply:
It's black formatted, so you would have to use the usual comment to deactivate black. But maybe itś better to not add the `python` then.